### PR TITLE
fix(coding-agent): apply agents-view renames optimistically instead of blocking on the daemon

### PR DESCRIPTION
**Pull request by snimu**
Fixes RES-1319

## Purpose

Renaming a session from the agents view (`space` → `/name <x>` or Ctrl+R) visibly froze the view: it sat on "Renaming agent..." until the daemon round-trip finished, and other daemon-backed interactions stalled behind it.

## Root cause (measured)

The freeze is daemon latency, not worker hydration: every rename — resident (`rename`) and saved (`rename_saved_session`) alike — asserts family-wide name availability via `createAgentFamilyCatalog`, which runs `SessionManager.listAll` (a full parse of every session file on disk) plus supervisor peer round-trips. Measured on a realistic roster (209 session files, 549 MB, warm cache): ~0.9–1.0 s for the scan alone per rename, more cold or with busy supervisor peers. Renaming an inactive session does NOT wake its worker (`rename_saved_session` appends a metadata entry to the file), so no hydration fix applies.

## Fix (view seam; the scan itself is catalog-cache territory, #2043/#2051)

The view never blocks on the rename round-trip:

- Submit applies the new name optimistically to the catalogs the reconcile reads (roster summaries and the saved catalog), rebuilds, and returns immediately — the row shows the new name at once and input stays live. The status shows `Renaming to <name>...`.
- The RPC completes in the background. Success confirms through the normal roster/saved refreshes and reports `Renamed to <name>`.
- Failure surfaces the error in the status line and triggers the same refreshes, which restore the previous name from daemon truth — no timers, no optimistic-state bookkeeping beyond the catalog rewrite.

## Tests

- Reworked the existing rename pin (fail-unfixed against main, where it blocks on a pending round-trip): submit resolves while the daemon promise is still pending with the optimistic name already applied and the composer disarmed; completion confirms via refreshes with the saved refetch gated on catalog load; a failing rename surfaces `Failed to rename agent: ...` and refetches to revert.
- The confirmRename capture pin gains the new collaborators; no new test files.

## LOC

Total src: **+60/−13 (net +47)**; tests: +223/−7 (net +216). Src is the non-blocking rename path: optimistic overlay with display-truth self-clearing at reconcile; no timers, no new state machines. +1 changelog fragment.

## Validation

- `npm run check` green (pre-commit hook).
- agents-view suites + 502 regressions green locally and in sandbox; fail-unfixed proven by reverting the src hunks (rename pin fails on the old blocking behavior).

---

> [!NOTE]
> **Medium Risk**
> Optimistic rename overlays and serialized daemon writes change session display and delete ordering; a failed RPC clears the overlay while a late success could briefly show a stale name until the next refresh.
> 
> **Overview**
> Fixes agents view **freezing on rename** by no longer awaiting the slow daemon rename RPC before updating the UI.
> 
> **`renameSession`** now applies the requested name immediately (`applyOptimisticSessionName` + `reconcileCatalogs`), shows `Renaming to …`, and returns while a **single per-session background writer** (`completeRename` on a **dedicated daemon client**) persists the name. Rapid renames update one pending entry and serialize writes so the **newest name wins**. **`reconcileCatalogs`** overlays pending names on live and saved catalogs via **`withPendingRenames`**, clearing the overlay only when catalog truth matches and no write is in flight (including edge cases like renaming back to a stale catalog name).
> 
> **Saved-session delete** now **awaits** any in-flight rename for that session before deleting, so a late rename write cannot recreate the file.
> 
> Tests cover non-blocking `/name`, overlay behavior across roster/saved refreshes, failed writes, dedicated-client lifetime, delete ordering, and no-op confirm when the name is unchanged. Minor formatting-only edits in `daemon-mode.ts`; changelog fragment added.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8771d5d146cb3197e30a070f9fd3f052476bd274. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Apply agents-view renames optimistically without blocking on daemon
> Changes `AgentsViewMode.renameSession` to update the displayed name immediately and write to the daemon in the background, so command submission no longer waits for the daemon round trip.
>
> - Adds per-session pending-rename state (`pendingRenames`, `PendingRename`) that tracks the newest requested name and a single serialized writer per session
> - `completeRename` uses a dedicated daemon client, loops when a newer name arrives mid-write, and removes the optimistic overlay only on failure when no newer request is pending
> - `reconcileCatalogs` and `withPendingRenames` overlay pending names over stale live and saved catalog truth, clearing the overlay only after the confirming catalog source reports the new name
> - `handleDeleteSelected` now awaits an in-flight rename writer before deleting a saved session
> - Behavioral Change: catalog refreshes no longer overwrite a pending rename with stale truth; the overlay persists until the owning catalog source confirms the new name or the write fails
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 8771d5d.</sup>
> 
>

### CODE REVIEW

File: packages/coding-agent/src/modes/agents-view/agents-view-mode.ts [hunk 1]
```diff
@@ -1667,37 +1667,63 @@ export class AgentsViewMode implements Component, Focusable {
...
+			entry.sessionId === summary.sessionId ? { ...entry, sessionName: name } : entry,
+		);
+		if (summary.sessionFile) {
+			const renamedPath = resolvePath(canonicalizePath(summary.sessionFile));
+			this.savedSessions = this.savedSessions.map((entry) =>
+				resolvePath(canonicalizePath(entry.path)) === renamedPath ? { ...entry, name } : entry,
+			);
+			this.persistentState.savedSessions = this.savedSessions;
+		}
+		this.reconcileCatalogs();
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/agents-view/agents-view-mode.ts:1674]: Fixed in 7d303c2b4. The optimistic name no longer lives as a one-off rewrite of `lastListedSummaries`/`savedSessions`: it is now a pending-rename overlay (sessionId -> name) applied on every catalog reconcile, so roster pushes, saved-catalog refreshes, and progressive scan streaming all keep the optimistic label while the RPC is in flight. The overlay is deleted when the RPC settles — success confirms and failure reverts through the normal refreshes. Pinned: a roster push mid-rename keeps the optimistic row name; completion clears the overlay (fails without it).

File: packages/coding-agent/src/modes/agents-view/agents-view-mode.ts [hunk 2]
```diff
@@ -1667,37 +1667,63 @@ export class AgentsViewMode implements Component, Focusable {
...
 			this.setStatusMessage(
 				isUnknownDaemonCommandError(error, "rename")
 					? "Failed to rename: the daemon is running an older build; restart the daemon and try again"
 					: formatError("Failed to rename agent", error),
 			);
-			return false;
 		}
+		// On success this confirms the optimistic name; on failure it reverts it.
+		await this.refreshSessions();
+		this.refreshSavedSessionsIfLoaded();
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/agents-view/agents-view-mode.ts:1720]: Fixed in 7d303c2b4. The optimistic name no longer lives as a one-off rewrite of `lastListedSummaries`/`savedSessions`: it is now a pending-rename overlay (sessionId -> name) applied on every catalog reconcile, so roster pushes, saved-catalog refreshes, and progressive scan streaming all keep the optimistic label while the RPC is in flight. The overlay is deleted when the RPC settles — success confirms and failure reverts through the normal refreshes. Pinned: a roster push mid-rename keeps the optimistic row name; completion clears the overlay (fails without it).

File: packages/coding-agent/src/modes/agents-view/agents-view-mode.ts [hunk 3]
```diff
@@ -1667,37 +1669,64 @@ export class AgentsViewMode implements Component, Focusable {
...
 			this.setStatusMessage(
 				isUnknownDaemonCommandError(error, "rename")
 					? "Failed to rename: the daemon is running an older build; restart the daemon and try again"
 					: formatError("Failed to rename agent", error),
 			);
-			return false;
 		}
+		// The overlay dies with the RPC; the refreshes then confirm the new name
+		// (success) or restore the previous one from daemon truth (failure).
+		this.pendingRenames.delete(summary.sessionId);
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/agents-view/agents-view-mode.ts]: Fixed in 58a8481f6. The overlay is no longer dropped at settle: on success the entry stays and the catalog reconcile self-clears it only when refreshed truth carries the overlay value — so a successful inactive rename holds the new name across the entire in-flight saved refetch (pinned with a deferred-refetch probe: rename settles, refetch pending, no old-name flash; refetch lands, overlay self-clears).

[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/agents-view/agents-view-mode.ts]: Fixed in 58a8481f6 by the same ownership rule: an overlay entry is cleared only by the actor whose value it holds. Success never deletes (the reconcile self-clears on truth match, so a stale first settle cannot kill a newer name — last writer wins); failure deletes only if the entry still holds its own name. Pinned: a second rename during the first one's flight keeps the second name through both settles.

File: packages/coding-agent/src/modes/agents-view/agents-view-mode.ts [hunk 4]
```diff
@@ -1667,37 +1669,79 @@ export class AgentsViewMode implements Component, Focusable {
...
+		if (this.pendingRenames.size === 0) return [...summaries];
+		return summaries.map((summary) => {
+			const name = this.pendingRenames.get(summary.sessionId);
+			if (name === undefined) return summary;
+			// Truth carrying the overlay value confirms the rename: the entry
+			// clears itself, so it can never drop early (before the saved refetch
+			// lands) nor outlive a newer rename's overlay (last writer wins).
+			if (summary.sessionName === name) {
+				this.pendingRenames.delete(summary.sessionId);
+				return summary;
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/agents-view/agents-view-mode.ts:1685]: Fixed in afb577e42 as the display-truth form of the invariant: rows read daemon truth first, so the saved-catalog branch now confirms (self-clears) an overlay entry only for saved-only sessions; a session that still carries a roster summary confirms exclusively from the roster. One-condition scoping, no new state. Pinned end-to-end: the progressive scan returns the renamed file while the roster summary is stale → the name never flashes and the overlay survives; the roster push lands → the overlay self-clears (fails on 16ef096b4).

File: packages/coding-agent/src/modes/agents-view/agents-view-mode.ts [hunk 5]
```diff
@@ -1654,27 +1661,87 @@ export class AgentsViewMode implements Component, Focusable {
...
+			return false;
+		}
+		const idle = this.pendingRenames.get(summary.sessionId)?.write === undefined;
+		this.applyOptimisticSessionName(summary, name);
+		this.setStatusMessage(`Renaming to ${name}...`);
+		// One writer per session: two concurrent writes can land in either order, so
+		// an older name could overwrite the newest one the user asked for.
+		if (idle) {
+			const pending = this.pendingRenames.get(summary.sessionId)!;
+			pending.write = this.completeRename(summary, pending);
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/agents-view/agents-view-mode.ts]: Fixed in 8771d5d14: `completeRename` now opens a dedicated client via `connectDedicatedClient()` before the write loop and closes it first in its `finally`, and `writeRename` sends both the `rename` RPC and `renameDaemonSavedSession` over that client, mirroring `createNewSession`. The exit-path `rosterClient.close()` (`daemon-client.ts:428`) rejects all pending requests, so the rename no longer rides the client that close hits.
Regression test "PR #2099: rename writes ride a dedicated client that outlives the shared roster client" fails on 5ecb72f1b (the rename command is observed on the shared client) and passes on 8771d5d14 with the shared client closed mid-write: the RPC lands on the dedicated client and completes, the row keeps the new name, and the dedicated client is closed after the loop.

*[written by prime-agent (RSI fleet), not human-reviewed]*

[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/agents-view/agents-view-mode.ts]: Fixed in 8771d5d14: `renameSession` re-fetches the pending entry after `applyOptimisticSessionName` instead of asserting it non-null. Confirming a name the catalog already carries is exactly the reconcile that `withPendingRenames` treats as confirmation — it deletes the overlay — so the rename is a settled no-op and no writer starts; nothing rejects.
Regression test "PR #2099: confirming a name the catalog already carries is a no-op, not a crash" reproduces the reported crash on 5ecb72f1b (promise rejected `TypeError: Cannot set properties of undefined (setting 'write')` at `renameSession`, plus the fire-and-forget unhandled rejection) and passes on 8771d5d14: resolves true, sends no rename command, leaves `pendingRenames` empty.

*[written by prime-agent (RSI fleet), not human-reviewed]*

File: packages/coding-agent/src/modes/agents-view/agents-view-mode.ts [hunk 6]
```diff
@@ -1654,27 +1661,87 @@ export class AgentsViewMode implements Component, Focusable {
...
+		let written: string | undefined;
+		try {
+			while (pending.name !== written && this.pendingRenames.get(summary.sessionId) === pending) {
+				const name = pending.name;
+				if (!(await this.writeRename(summary, name))) {
+					// Only the failing owner clears its entry; success waits for truth to match.
+					if (this.pendingRenames.get(summary.sessionId) === pending) {
+						this.pendingRenames.delete(summary.sessionId);
+					}
+					return;
```
[the remark this answers is not in the document]
Comment (snimu) [reply packages/coding-agent/src/modes/agents-view/agents-view-mode.ts:1738]: Fixed in 8771d5d14: the failure branch of `completeRename` now checks `pending.name !== name` before clearing the entry and continues the loop, so a newer name requested during the failed write still gets its turn. The loop condition still guards ownership and a failure on the NEWEST name clears the entry as before.
Regression test "PR #2099: a failed rename write still gives the newer name its turn" fails on 5ecb72f1b (only "First Name" is ever sent; "Second Name" is dropped by the failed write) and passes on 8771d5d14: after the first write fails, "Second Name" is sent and the row shows it.

*[written by prime-agent (RSI fleet), not human-reviewed]*
